app/models/story_parser.rb

```diff
@@ -545,10 +545,25 @@ def download_chaptered_from_efiction(location)
   def parse_common(story, location = nil, encoding = nil, detect_tags = true)
     work_params = { title: "Untitled Imported Work", chapter_attributes: { content: "" } }
 
+    # check if the given encoding is valid, otherwise fallback to nil
+    begin
+      Encoding.find(encoding) if encoding
+    rescue ArgumentError
+      encoding = nil
+    end
+
     # Encode as HTML - the dummy "foo" tag will be stripped out by the sanitizer but forces Nokogiri to
     # preserve line breaks in plain text documents
     # Rescue all errors as Nokogiri complains about things the sanitizer will fix later
-    @doc = Nokogiri::HTML.parse(story.prepend("<foo/>"), nil, encoding) rescue ""
+    story.prepend("<foo><foo/>")
+    @doc = begin
+             # Using Nokogiri::HTML4 here for its better encoding handling.
```

Comment:
Thank you for the investigation.

The tests were introduced in https://github.com/otwcode/otwarchive/commit/0e59af31c028404c932acbafffb5a8a94dde7590 / [AO3-4628](https://otwarchive.atlassian.net/browse/AO3-4628) and haven't been changed since then. A straightforward implementation of using the HTML5 parser would break one of the tests, if I recall it correctly.

It seems like the intention was to prevent crashing against exoteric encodings. I don't know if that included outright wrong encodings, but I think it's good to prepare for that possibility, too. I believe [Ruby 3](https://otwarchive.atlassian.net/browse/AO3-6409) changed a lot about string encodings, so some updating might be necessary here because of that, too.

I think we could do better with HTML5 than I initially thought with a small change. In particular, we can rescue from `Encoding::UndefinedConversionError` and specify `binary` or `ASCII-8bit`, instead of leaving the encoding `nil`. It might not get you correct conversion, but it looks like it would at least avoid crashing. (I think I will add a commit that does this later.)

Reply:
Perhaps I should ask this first: given input with wrong encoding (a UTF-8 string but Windows-1252 is specified, for example), which do we prefer? This is mostly about importing.

1. Go on with potentially [malformed](https://en.wikipedia.org/wiki/Mojibake) content - I think this is what we do currently 
2. Abort with an error
3. Guess and recover the correct encoding statistically (ignoring the specified encoding) - good user experience but, requires significantly more code